xarray/testing.py

```diff
@@ -0,0 +1,143 @@
+        arr1 = _decode_string_data(arr1)
+        arr2 = _decode_string_data(arr2)
+    exact_dtypes = ['M', 'm', 'O', 'U']
+    if any(arr.dtype.kind in exact_dtypes for arr in [arr1, arr2]):
+        return ops.array_equiv(arr1, arr2)
+    else:
+        return ops.allclose_or_equiv(arr1, arr2, rtol=rtol, atol=atol)
+
+
+def assert_equal(a, b):
+    """Like :py:func:`numpy.testing.assert_array_equal`, but for xarray
+    objects.
+
+    Raises an AssertionError if two objects are not equal. This will match
+    data values and coordinates, but not names or attributes (except for
+    Dataset objects for which the variable names must match).
+
+    Parameters
+    ----------
+    a : xarray.Dataset or xarray.DataArray
+        The first object to compare
+    b : xarray.Dataset or xarray.DataArray
+        The second object to compare
+
+    See also
+    --------
+    assert_identical, assert_allclose, xarray.Dataset.equals,
+    xarray.DataArray.equals
+    """
+    import xarray as xr
+    ___tracebackhide__ = True  # noqa: F841
+    assert type(a) == type(b)
+    if isinstance(a, (xr.Variable, xr.DataArray, xr.Dataset)):
+        assert a.equals(b), '{}\n{}'.format(a, b)
+    else:
+        raise TypeError('{} not supported by assertion comparison'
+                        .format(type(a)))
+
+
+def assert_identical(a, b):
+    """Like :py:func:`xarray.testing.assert_equal`, but also matches the
+    objects' names and attributes.
+
+    Raises an AssertionError if two objects are not identical.
+
+    Parameters
+    ----------
+    a : xarray.Dataset or xarray.DataArray
+        The first object to compare
+    b : xarray.Dataset or xarray.DataArray
+        The second object to compare
+
+    See also
+    --------
+    assert_equal, assert_allclose, xarray.Dataset.equals,
+    xarray.DataArray.equals
+    """
+    import xarray as xr
+    ___tracebackhide__ = True  # noqa: F841
+    assert type(a) == type(b)
+    if isinstance(a, xr.DataArray):
+        assert a.name == b.name
+        assert_identical(a._to_temp_dataset(), b._to_temp_dataset())
+    elif isinstance(a, (xr.Dataset, xr.Variable)):
+        assert a.identical(b), '{}\n{}'.format(a, b)
+    else:
+        raise TypeError('{} not supported by assertion comparison'
+                        .format(type(a)))
+
+
+def assert_allclose(a, b, rtol=1e-05, atol=1e-08, decode_bytes=True):
+    """Like :py:func:`numpy.testing.assert_allclose`, but for xarray objects.
+
+    Raises an AssertionError if two objects are not equal up to desired
+    tolerance.
+
+    Parameters
+    ----------
+    a : xarray.Dataset or xarray.DataArray
+        The first object to compare
+    b : xarray.Dataset or xarray.DataArray
+        The second object to compare
+    rtol : float, optional
+        Relative tolerance
+    atol : float, optional
+        Absolute tolerance
+    decode_bytes : bool, optional
+        Whether byte types should be decoded to a string or not
```

Comment:
Maybe:

    Whether byte dtypes should be decoded to strings as UTF-8 or not.
    This is useful for testing serialization methods on Python 3 that return
    saved strings as bytes.